src/ir/module-utils.h

```diff
@@ -106,6 +106,24 @@ inline Table* copyTable(const Table* table, Module& out) {
   return out.addTable(std::move(ret));
 }
 
+inline DataSegment* copyDataSegment(const DataSegment* segment, Module& out) {
+  auto copy = [&](std::unique_ptr<DataSegment>&& ret) {
+    ret->name = segment->name;
+    ret->hasExplicitName = segment->hasExplicitName;
+    ret->isPassive = segment->isPassive;
+    ret->data = segment->data;
+
+    return out.addDataSegment(std::move(ret));
+  };
+
+  if (segment->isPassive) {
+    return copy(std::make_unique<DataSegment>());
```

Comment:
It would be slightly simpler to unconditionally create and fill out the new `DataSegment` then conditionally set its `offset`.